src/kernels/gibbskernel.jl

```diff
@@ -0,0 +1,37 @@
+"""
+
+    GibbsKernel(x, y)
+
+# Definition
+
+The Gibbs kernel is non-stationary generalisation of the Squared-Exponential
+kernel. The lengthscale parameter ``\\ell`` becomes a function of
+position ``\\ell(x)``.
+
+``\\ell(x) = 1.`` then you recover the standard Squared-Exponential kernel
+with constant lengthscale.
+
+```math
+k(x, y) = \\sqrt{ \\left( \\frac{2 \\ell(x) \\ell(y)}{\\ell(x)^2 + \\ell(y)^2} \\right) } \\quad \\rm{exp} \\left( - \\frac{(x - y)^2}{\\ell(x)^2 + \\ell(y)^2} \\right)
+```
+
+[1] - Mark N. Gibbs. "Bayesian Gaussian Processes for Regression and Classication."
+    PhD thesis, 1997
+
+[2] - Christopher J. Paciorek and Mark J. Schervish. "Nonstationary Covariance
+    Functions for Gaussian Process Regression". NEURIPS, 2003
+
+[3] - Sami Remes, Markus Heinonen, Samuel Kaski.
+    "Non-Stationary Spectral Kernels". arXiV:1705.08736, 2017
+
+[4] - Sami Remes, Markus Heinonen, Samuel Kaski.
+    "Neural Non-Stationary Spectral Kernel". arXiv:1811.10978, 2018
+"""
+
+#function (k::GibbsKernel)(x, y, ell)
+function GibbsKernel(x::Any, y::Any, ell::Function)
+    fac = ell(x)^2 + ell(y)^2
+    term1 = 2.0 * ell(x) * ell(y) / fac
+    term2 = - (x-y)^2 / fac
+    return term1 * exp(term2)
+end
```

Comment:
```suggestion
struct GibbsKernel{T} <: Kernel
    ell::T
end

function (k::GibbsKernel)(x, y)
    fac = k.ell(x)^2 + k.ell(y)^2
    term1 = 2 * k.ell(x) * k.ell(y) / fac
    term2 = -(x-y)^2 / fac
    return term1 * exp(term2)
end
```
I think all of the functionality is here. The main change this suggestion makes is to create a `Kernel` called `GibbsKernel` and define a way to evaluate it. This is needed to ensure that the kernel behaves in the same way as our other kernels.

Minor changes -- these are new-to-Julia things that are more general than this PR:
1. remove `::Any` restrictions on `x` and `y`. `Any` is implicitly assumed if no type constraint is specified.
2. `2.0` -> `2`. This will generally interact better with Julia promotion mechanisms. For example, if `k.ell(x) * k.ell(y) / fac` is a `Float32` then probably what we want is for `term1` to be a `Float32`. Multiplying by `2.0` will produce a `Float64`, because `Float64 * Float32 -> Float64`, while multiplying by `2` will give a `Float32` as desired.

In the tests, you'll now be able to write things like
```julia
k_gibbs = GibbsKernel(ell)
k_gibbs(x, y)
```
etc.

Reply:
Another small thing:
Assuming that `ell` could be expensive, it might worth precomputing them first and then using them.

Reply:
Agreed -- I think this should be done in a follow up PR as an optimisation for `kernelmatrix` etc.

Reply:
I think it would be good to allow to specify a distance measure, as in the case of `SqExponentialKernel`. Maybe one could (should?) also implement `(::GibbsKernel)(x, y)` by constructing a corresponding scaled `SqExponentialKernel` with the correct lengthscale and evaluate it to make use of the optimizations for lengthscales and certain distance measures. In any case, we should not use `(x - y)^2` since it will fail for non-scalar inputs.

Reply:
Agreed -- could we fix this in a follow-up PR though? Perhaps we just restrict the input types to `Real` in this PR?

Reply:
It would be a breaking change to add a new field for the distance measure (since it implies an additional type parameter and hence can break dispatches).

Reply:
Good point. Could you propose a suggestion that extends my one?

Reply:
Something like
```suggest
struct GibbsKernel{T,M} <: Kernel
    l::T
    metric::M
end

GibbsKernel(; l, metric=Euclidean()) = GibbsKernel(l, metric)

function (k::GibbsKernel)(x, y)
    l = k.l
    lx = l(x)
    ly = l(y)
    lengthscale = (lx^2 + ly^2) / 2
    kernel = (lx * ly / lengthscale) * with_lengthscale(SqExponentialKernel(k.metric), lengthscale)
    return kernel(x, y)
end
```
I think though we should use something more descriptive than `l` or `ell` though.

Reply:
I don't think that would be correct though!  The square applied on the `lengthscale` terms comes from the `SqEuclidean` norm. For different norms, I believe `lengthscale` would be different.

Reply:
Maybe, I did not study the original issue or the motivation for this particular choice 😄 It is always a squared exponential kernel though, just with a different metric, so the overall mathematical structure is the same. If necessary, we could make the computation of the variance and lengthscale also dependent on the metric.

Reply:
All I am saying it that we should probably not allow for other metrics than `SqEuclidean` for now.

Reply:
We use `Euclidean` as default. The squared metric, scaled with 1/2, is used and fixed in the definition of `SqExponentialKernel`. We always have a squared metric (which I referred to as same mathematical structure), and therefore I am less worried about the use of other metrics.

Reply:
Following on from Theo's point, a potential issue with using other metrics is that the Gibbs kernel is defined using some kind of integral against a white noise process (I forget the precise details), so it's possible that the right thing to do under metrics other than the Euclidean has some parametric form other than the one above.

Reply:
I'll read up on the details later today and check if it makes sense to allow to use a general metric. We could of course also just restrict the implementation to `(::GibbsKernel{<:Any,<:Euclidean})(x, y)` for now (this would avoid breaking changes in the future) or remove the `metric` field (if we think we don't want/need it and/or don't care about breaking changes).

In any case, my suggestion above was incorrect, it should be of course something like
```suggestion
struct GibbsKernel{T,M} <: Kernel
    l::T
    metric::M
end

GibbsKernel(; l, metric=Euclidean()) = GibbsKernel(l, metric)

function (k::GibbsKernel)(x, y)
    l = k.l
    lx = l(x)
    ly = l(y)
    lengthscale = hypot(lx, ly)
    kernel = (lx * ly / lengthscale^2) * with_lengthscale(SqExponentialKernel(k.metric), lengthscale)
    return kernel(x, y)
end
```

Reply:
> remove the metric field (if we think we don't want/need it and/or don't care about breaking changes).

Hmmm honestly I'm inclined to think that this is the case. If we need to make a breaking change at some point to support different metrics, so be it 🤷 .

I do agree that we ought to ensure that we either restrict the argument type to `Real`s, or generalise the implementation to vectors of reals. I'm inclined to think that former is better for the sake of getting this in quickly, so something like
```suggestion
struct GibbsKernel{T} <: Kernel
    ell::T
end

function (k::GibbsKernel)(x::Real, y::Real)
    fac = k.ell(x)^2 + k.ell(y)^2
    term1 = 2 * k.ell(x) * k.ell(y) / fac
    term2 = -(x-y)^2 / fac
    return term1 * exp(term2)
end
```

Reply:
I really think we shouls use `SqExponentialKernel` already in the first implementation since I think we want to use it at least in the vector case at some point - but then it would be another breaking change to switch to it since it uses a slightly different parameterization (factor 1/2 in the exponential).

I agree, probably it's easiest to not include the `metric` for now. Therefore I'd suggest, with a better name for `l`/`ell`:
```suggestion
struct GibbsKernel{T} <: Kernel
    l::T
end

GibbsKernel(; l) = GibbsKernel(l)

function (k::GibbsKernel)(x, y)
    l = k.l
    lx = l(x)
    ly = l(y)
    lengthscale = hypot(lx, ly)
    kernel = (sqrt(2 * lx * ly) / lengthscale) * with_lengthscale(SqExponentialKernel(), lengthscale)
    return kernel(x, y)
end
```
(BTW it seems `term1` is incorrect in all suggestions and the original implementation and does not correspond to the formula in the issue)

Reply:
Thanks for such a lively discussion everyone!

@devmotion I just implemented your suggested and thanks for pointing out the error in `term1`!

I changed your `l` to `lengthscale` to hopefully make it obvious what it is and solve the `ell`/`l` issue.

I'm not entirely confident in my unit test either so will have to double check that too. I think it's because of the `1/2` in the `SqExponentialKernel` but I'll double check.

Nice idea about using the `with_lengthscale`.